tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,338 @@
+            model.compile(loss = tf.keras.losses.BinaryCrossentropy, optimizer = opt)
+            model.fit(x, y)
+
+        Arguments
+            base_optimizer: A class that inherits from tf.keras.optimizers.Optimizer. Do not
+                pass an instance of the class.
+
+            model: tf.keras.Model, The model to be used for discriminative learning.
+                It should have at least 1 layer with the attribute lr_mult. The lr_mult should
+                be set to a value not equal to 1. Otherwise, you will have the exact same
+                result as not using discriminative learning.
+
+            learning_rate: float, the learning rate for the model
+
+            verbose: Bool, to generate a report on how many parameters are affected
+
+            *args: Args to pass to the base optimizer
+
+            **kwargs: Kwargs to pass to the base optimizer
+
+        Returns
+            Optimizer - A keras optimizer to use with model.compile
+
+        References
+            - [Universal Language Model Fine-tuning for Text Classification](https://arxiv.org/pdf/1801.06146.pdf)
+        """
+
+        assert issubclass(
+            base_optimizer, tf.keras.optimizers.Optimizer
+        ), "Base optimizer must be a class that inherits from tf.keras.optimizers.Optimizer"
+
+        # assume that users will follow the general guidelines and init thier opts within a dist scope.
+        if tf.distribute.has_strategy():
+            logging.warning(
+                """The discriminative layer optimizer may not behave as expected
+            when using a distribution strategy.
+            """
+            )
+
+        super().__init__(lr=learning_rate, name=name, *args, **kwargs)
+
+        DiscriminativeModelManager._prepare_model(model, verbose=verbose)
+
+        self.opt_class = base_optimizer
+        self.learning_rate = learning_rate
+        self.kwargs = kwargs
+
+        # Find unique lr_mult.
+        variable_groups = {var.lr_mult: None for var in model.trainable_variables}
+
+        self.optimizer_group = []
+
+        for lr_mult in variable_groups.keys():
+            opt = self.opt_class(learning_rate=learning_rate * lr_mult, **kwargs)
+            opt.lr_mult = lr_mult
+            self.optimizer_group.append(opt)
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        """Allocates gradients to each optimizer based on the variable's learning rate multiplier
+        then applies the gradients. In graph mode, it returns 1 operation per optimizer.
+        Please use the model.fit method instead of accessing this directly.
+        """
+
+        # Create gradvar buckets for each opt.
+        gvdict = {}
+        for opt in self.optimizer_group:
+            gvdict[opt.lr_mult] = []
+
+        # Load the gradvars into the appropriate bucket.
+        for grad, var in tuple(grads_and_vars):
+            gvdict[var.lr_mult].append((grad, var))
```

Comment:
we could store a "master" list of variables on initialization and always assign variables based on that list, instead of the current state of the variable? 

This optimizer is intended to be used within a model.fit command, so I guess the user could create a callback that changes the lr_mult of the variable, but that isn't supported by the current implementation.

Reply:
>  but that isn't supported by the current implementation.

Mention this clearly in the docstring. 

Also, the current implementation is fine, but you should error out if `lr_mult` has been changed during training.

Reply:
I'll make that change